gp-okta.py

```diff
@@ -385,67 +536,200 @@ def okta_redirect(conf, s, session_token, redirect_url):
+		verify = conf.get('openconnect_certs').name
+	h, c = send_req(conf, s, 'okta redirect form (2)', url, data,
+		expected_url='https://{0}'.format(gateway), verify=verify)
+	saml_username = h.get('saml-username', '').strip()
+	if len(saml_username) == 0:
+		err('saml-username empty')
+	prelogin_cookie = h.get('prelogin-cookie', '').strip()
+	if len(prelogin_cookie) == 0:
+		err('prelogin-cookie empty')
 
+	return saml_username, prelogin_cookie
 
 def main():
-	if len(sys.argv) < 2:
-		print('usage: {0} <conf>'.format(sys.argv[0]))
-		sys.exit(1)
-	conf = load_conf(sys.argv[1])
-	
+
+	parser = argparse.ArgumentParser(description="""
+	This is an OpenConnect wrapper script that automates connecting to a
+	PaloAlto Networks GlobalProtect VPN using Okta 2FA.""")
+
+	parser.add_argument('conf_file',
+		help='e.g. ~/.config/gp-okta.conf')
+	parser.add_argument('--gpg-decrypt', action='store_true',
+		help='use gpg and settings from gpg-home to decrypt gpg encrypted conf_file')
+	parser.add_argument('--gpg-home', default=os.path.expanduser('~/.gnupg'))
+	parser.add_argument('--show-list-of-gateways', default=False, action='store_true',
+		help='get list of gateways from portal')
+	parser.add_argument('--quiet', default=False, action='store_true',
+		help='disable verbose logging')
+	args = parser.parse_args()
+
+	global quiet
+	quiet = args.quiet
+
+	assert os.path.exists(args.conf_file)
+	assert not args.gpg_decrypt or os.path.isdir(args.gpg_home)
+
+	config_contents = ''
+	with io.open(args.conf_file, 'rb') as fp:
+		config_contents = fp.read()
+
+	if args.conf_file.endswith('.gpg') and not args.gpg_decrypt:
+		err('conf file looks like gpg encrypted. Did you forget the --gpg-decrypt?')
+
+	if args.gpg_decrypt:
+		if not have_gnupg:
+			err('Need gnupg package for reading gnupg encrypted files. Consider doing `pip install python-gnupg` (or similar)')
+		gpg = gnupg.GPG(gnupghome=args.gpg_home)
+		decrypted_contents = gpg.decrypt(config_contents)
+
+		if not decrypted_contents.ok:
+			print('[ERROR] failed to decrypt config file:', file=sys.stderr)
+			print('[ERROR]     status: {}'.format(decrypted_contents.status), file=sys.stderr)
+			print('[ERROR]     error: {}'.format(decrypted_contents.stderr), file=sys.stderr)
+			sys.exit(1)
+
+		config_contents = decrypted_contents.data
+
+	conf = load_conf(config_contents)
+
 	s = requests.Session()
 	s.headers['User-Agent'] = 'PAN GlobalProtect'
-	saml_xml = paloalto_prelogin(conf, s)
+
+	if conf.get('client_cert'):
+		s.cert = conf.get('client_cert')
+
+	if args.show_list_of_gateways:
+		userauthcookie, gateways = paloalto_getconfig(conf, s)
+		print("Gateways:")
+		for gateway in sorted(gateways):
+			print(" ", gateway)
+		sys.exit(0)
+
+	another_dance = conf.get('another_dance', '').lower() in ['1', 'true']
+
+	if conf.get('gateway'):
+		gateway = conf.get('gateway').strip()
+
+	if gateway and not another_dance:
```

Comment:
If gateway isn't defined in the conf, then this will produce an error in Python 3.7.5 at least (gateway used before assignment)

Reply:
Thank you @cardonator. Fixed.